### DISCUSSION TITLE
A Way to Change the Volumes of All the Players at One Time With a Keyboard Shortcut or With EasyEffects Sink?

### DISCUSSION THREAD

**Discussion by lisphm** [wwmm/easyeffects#2854] [Q&A] [answered]
My laptop has a subpar EQ for its speakers, so I'm using this setup.
![Screenshot from 2024-01-12 18-06-56](https://github.com/wwmm/easyeffects/assets/116984726/054e6617-2627-4717-b774-c57550e0c631)
I need to boost the maximum volume, but without a Limiter, the sound becomes very distorted. The only problem is that the Limiter worsens the sound quality when things are very loud, even with the Auto Leveling. I'm stuck between two choices: have the speakers be too quiet so it never reaches the Limiter, or have the sound quality always be worse. Changing the system volume only changes the volume _after_ the Limiter has been applied. However, I want some way to change the volume _before_ the Limiter is applied so that I only lose sound quality at the highest volume levels.

I've been thinking if there is a way to change the Output Volume value in the Loudness component based on EasyEffects Sink, or set it to a variable that I can easily change with a keyboard shortcut.
![Screenshot from 2024-01-12 18-10-00](https://github.com/wwmm/easyeffects/assets/116984726/a024e9d7-bffe-442f-962c-052c6113527e)
Thanks for your help.

_Oh and also, EasyEffects Sink does absolutely nothing, even if I set my sound output device to it. Is this intended?_

**Comment by lisphm** [comment]
If all else fails, I can just change each Player's volume individually. Or better yet, find a way to scale all of the volumes of the Players based on one value, which I suppose is the same thing as the original question.
New Title:
A Way to Change the Volumes of All the Players at One Time With a Keyboard Shortcut or With EasyEffects Sink?

**Reply by wwmm** [reply]
>Oh and also, EasyEffects Sink does absolutely nothing, even if I set my sound output device to it. Is this intended?

By default PipeWire does not enable volume and mute controls for virtual sinks. It is possible to change its configuration files to force it to do otherwise. But honestly it is better to not do it and keep controlling your soundcard device volume. 

>I need to boost the maximum volume, but without a Limiter, the sound becomes very distorted.

If you amplify the signal beyond `0 dB` you will most likely have distortion. There is no way around it. The level meter in EasyEffects bottom panel will warn you when this happens.

>I want some way to change the volume before the Limiter is applied so that I only lose sound quality at the highest volume levels.

It is not possible to amplify the level above `0 dB` without the limiter taking action for force the level down at all costs. It is its function after all.

Show a print of EasyEffects window when the audio level is "too low". Unless there is an output gain control somewhere decreasing the level there is nothing you can do. You may have already reached the limitations of your laptop speakers.

**Reply by lisphm** [reply]
> Show a print of EasyEffects window when the audio level is "too low".

Removed the Loudness and set the output volume of the EQ back to `0 dB`.
![Screenshot from 2024-01-12 22-25-13](https://github.com/wwmm/easyeffects/assets/116984726/40bc0d85-699f-4195-ad43-d00d824ea1e2)
All the values in the EQ are below or at `0 dB`, which is why it is quieter than desired.

However, I am referring to a simple and convenient way for the `Output Volume` to change the volume _before_ any sound effects are applied.
![Screenshot from 2024-01-12 22-47-34](https://github.com/wwmm/easyeffects/assets/116984726/fb61f285-72d5-4ecc-aa13-b7c91ebace05)
As of now, changing the `Output Volume` does not change the sound that the Equalizer receives.
I want a way to change what the Equalizer receives that is as easy as changing the `Output Volume`.

**Reply by wwmm** [reply, marked as answer]
>As of now, changing the Output Volume does not change the sound that the Equalizer receives.
I want a way to change what the Equalizer receives that is as easy as changing the Output Volume.

If you set `monitor.channel-volumes = true` somewhere in PipeWire's configuration file our sink volume will be enabled and you will get what you want. But I am not sure where in its configuration file this setting should be put. So it is worth to ask for this there or take a look at their wiki pages.

That being said I still do not see the reason why the volume has to be controlled before the equalizer. As far as the plugin is concerned it does not need this.

**Reply by lisphm** [reply]
OK, thanks.


The reason is because of the limiter. Since the volume is controlled after the limiter, it means that most of the time, the volume gets high enough to reach the limiter. But controlling the volume anywhere before the limiter means I can reach the limiter only when I need to.

This is an exaggerated example of what I'm talking about. The Output Volume (Speakers) is set to mute, but EasyEffects sees the same volume. This means that whatever I do to the output volume, the limiter will still act up.
![Screenshot from 2024-01-13 10-23-15](https://github.com/wwmm/easyeffects/assets/116984726/27857849-a5f2-4a03-9152-b41967dcc597)

However, changing the loudness before the limiter means that the limiter only acts when I need it to, saving sound quality.
![Screenshot from 2024-01-13 10-22-10](https://github.com/wwmm/easyeffects/assets/116984726/e18d7b5e-d97f-4620-b2a1-055c2c114465)

**Reply by wwmm** [reply]
Depending on how well the `autogain` plugin works for your audio content it will be an easier and automated solution to handle the level of the signal that gets before the limiter. And considering your concerns about the audio level the autogain would also probably allow you to not set the equalizer input gain so high.

**Comment by pointful** [comment]
For anyone else looking to achieve this configuration for similar reasons, I was able to make this work on Ubuntu 26.04 by creating `~/.config/pipewire/pipewire.conf.d/60-easyeffects.conf` with the following:

```
node.rules = [
  {
    matches = [
      {
        node.name = "easyeffects_sink"
      }
    ],
    actions = {
      update-props = {
        monitor.channel-volumes = true
      }
    }
  }
]
```

And then using `systemctl --user restart wireplumber pipewire` to reload.

And a comment overall:

Knowing that having this as the default previously created a lot of tech support issues, how about making it configurable via settings somewhere in the Easy Effect UI? That would be a bit easier to manage/maintain than having to set up the above.

**Reply by wwmm** [reply]
>Knowing that having this as the default previously created a lot of tech support issues, how about making it configurable via settings somewhere in the Easy Effect UI? That would be a bit easier to manage/maintain than having to set up the above.

It is technically possible. Maybe the app will have to be restarted for this to take effect. But it will probably work. I'm only afraid this will reinforce bad procedures in people's mind. In the past when people wanted these controls the reason was they were setting our virtual devices as their system default devices. Something that should never be done.

**Reply by gitvalds** [reply]
Thank you very much @pointful,
 Your solution works for me too, on Fedora 44.
 I ran all the commands as normal user (not root).

**Comment by pointful** [comment]
Thanks, it was just a suggestion in case more people might want this. Out of curiousity: Whty do you say it should never be used as the system default device? I totally get how it might be confusing/problematic for most users, but if someone knows what they are doing I could see it being useful.

Here is my specific reason for wanting it this way: I have a set of Audioengine A2+ speakers I am using at my desktop. I want to connect them via USB as the audio quality is better that way, but the hardware volume control I get from them to the desktop has a really weird volume curve. There is no sound at all until I get to 25% and then the lowest possible setting is louder than I sometimes want it. The volume also increases in big steps rather than a smooth curve. (I spent some time trying to fix this behavior directly but never made much progress.) So, I instead end up adjusting the volume of individuals streams, but this is tedious on daily basis. I was already using EasyEffects in the intended way for access to all the various filters, especially for listening to music.  What I am trying out now is setting the speakers to maximum comfortable volume for me, and then using the EasyEffects Sink as my main volume control and so I want to make it the system default device so that I can use my keyboard volume keys adjust overall volume up and down.

Anyway, I love using EasyEffects -- thanks so much for all the hard work on it -- it is a great tool! This new setup seems to be working for me for the moment as long as I remember what I am doing. :) Totally open to other suggestions on a better configuration to work around the wonky volume curve I currently get from my speakers. Thanks again!

**Reply by wwmm** [reply]
>Why do you say it should never be used as the system default device?

There is a discussion about that at #1836. But long story short EasyEffects is designed assuming your hardware stays as the system default device. If you set its virtual devices as default there is a chance easyeffects won't know what to use for sink and source. It will depend on whether pipewire send us the system default device information before or after our devices are loaded. Usually the event about the system default comes before our virtual devices are loaded and easyeffects gets the correct information it needs. But there is no guarantee.

As it makes no sense for easyeffects to play audio to itself or record silence from itself it can't use its own devices. In simple systems that have only one soundcard or microphone we could workaround that. But when there are more than one microphone or soundcard things get complicated. Trying to guess the device we should be using without relying in the default device information or manual selection done by the user becomes problematic. The easiest thing to do by far is requiring that the user hardware stays the default. And it is what most people do anyway.

>instead end up adjusting the volume of individuals streams

What I do in my computer is having all streams to 100% and adjusting my soundcard volume. The fact I usually use the autogain helps when apps start in a different volume for some reason. But usually they reopen at 100% volume.

**Comment by pointful** [comment]
Thanks for the summary and the pointer to the detailed discussion -- makes sense in the general case.

In my specific case where the soundcard volume curve of my speakers is problematic, it is working well to make the EasyEffects sink the main volume control for the moment.

Thanks again!
